Source/Core/Core/PowerPC/SignatureDB.h

```diff
@@ -5,40 +5,56 @@
 #pragma once
 
 #include <map>
+#include <memory>
 #include <string>
 
 #include "Common/CommonTypes.h"
 
 // You're not meant to keep around SignatureDB objects persistently. Use 'em, throw them away.
 
 class PPCSymbolDB;
+class SignatureDBFormatHandler;
 
 class SignatureDB
 {
+public:
   struct DBFunc
   {
-    std::string name;
     u32 size;
+    std::string name;
+    std::string object_name;
+    std::string object_location;
     DBFunc() : size(0) {}
   };
+  using FuncDB = std::map<u32, DBFunc>;
 
-  // Map from signature to function. We store the DB in this map because it optimizes the
-  // most common operation - lookup. We don't care about ordering anyway.
-  typedef std::map<u32, DBFunc> FuncDB;
-  FuncDB database;
-
-public:
   // Returns the hash.
   u32 Add(u32 startAddr, u32 size, const std::string& name);
 
-  bool Load(const std::string&
-                filename);  // Does not clear. Remember to clear first if that's what you want.
-  bool Save(const std::string& filename);
+  // Does not clear. Remember to clear first if that's what you want.
+  bool Load(const std::string& file_path);
+  bool Save(const std::string& file_path);
   void Clear();
   void List();
 
   void Initialize(PPCSymbolDB* func_db, const std::string& prefix = "");
   void Apply(PPCSymbolDB* func_db);
 
   static u32 ComputeCodeChecksum(u32 offsetStart, u32 offsetEnd);
+
+private:
+  std::unique_ptr<SignatureDBFormatHandler> GetFormatHandler(const std::string& file_path);
+  // Map from signature to function. We store the DB in this map because it optimizes the
+  // most common operation - lookup. We don't care about ordering anyway.
+  FuncDB m_database;
+};
+
+class SignatureDBFormatHandler
+{
+public:
+  // Implement in cpp file, to prevent the vtable being placed
+  // in every translation unit that uses this class.
+  virtual ~SignatureDBFormatHandler() = default;
```

Comment:
The above comment was meant to indicate that you should place the destructor implementation for SignatureDBFormatHandler into the cpp file so that it doesn't generate out-of-line vtables. It wasn't meant to be copied verbatim.